src/connectors/query-rules/connectQueryRules.ts

```diff
@@ -0,0 +1,237 @@
+  return ruleContexts;
+}
+
+function prepareRuleContexts({
+  initialRuleContexts,
+  newRuleContexts,
+  transformRuleContexts,
+}: {
+  initialRuleContexts: string[];
+  newRuleContexts: string[];
+  transformRuleContexts: ParamTransformRuleContexts;
+}) {
+  const allRuleContexts = [...initialRuleContexts, ...newRuleContexts];
+
+  warning(
+    allRuleContexts.length <= 10,
+    `
+The maximum number of \`ruleContexts\` is 10. They have been sliced to that limit.
+Consider using \`transformRuleContexts\` to minimize the number of rules sent to Algolia.
+`
+  );
+
+  const ruleContexts = transformRuleContexts(allRuleContexts).slice(0, 10);
+
+  return ruleContexts;
+}
+
+function applyRuleContexts({ helper, ruleContexts }) {
+  const previousRuleContexts = helper.getQueryParameter('ruleContexts');
+
+  if (!isEqual(previousRuleContexts, ruleContexts)) {
+    helper.setQueryParameter('ruleContexts', ruleContexts).search();
+  }
+}
+
+const connectQueryRules: QueryRulesConnector = (render, unmount = noop) => {
+  checkRendering(render, withUsage());
+
+  // We store the initial rule contexts applied before creating the widget
+  // so that we do not override them with the rules created from `trackedFilters`.
+  let initialRuleContexts: string[] = [];
+
+  // We track the added rule contexts to remove them when the widget is disposed.
+  let addedRuleContexts: string[] = [];
+
+  return widgetParams => {
+    const {
+      trackedFilters = {} as ParamTrackedFilters,
+      transformRuleContexts = (rules => rules) as ParamTransformRuleContexts,
+      transformItems = (items => items) as ParamTransformItems,
+    } = widgetParams || {};
+
+    Object.keys(trackedFilters).forEach(facetName => {
+      if (typeof trackedFilters[facetName] !== 'function') {
+        throw new Error(
+          withUsage(
+            `'The "${facetName}" filter value in the \`trackedFilters\` option expects a function.`
+          )
+        );
+      }
+    });
+
+    return {
+      init({ helper, state, instantSearchInstance }) {
+        if (Object.keys(trackedFilters).length > 0) {
+          initialRuleContexts = state.ruleContexts || [];
+
+          // The helper's method `getQueryParameter` doesn't work on unset attributes.
+          // We need to set `ruleContexts` to a default value before retrieving it.
+          if (initialRuleContexts.length === 0) {
+            helper.setQueryParameter('ruleContexts', initialRuleContexts);
+          }
+
+          const newRuleContexts = getRuleContextsFromTrackedFilters({
+            helper,
+            trackedFilters,
+          });
+          addedRuleContexts = newRuleContexts;
+          const ruleContexts = prepareRuleContexts({
+            initialRuleContexts,
+            newRuleContexts,
+            transformRuleContexts,
+          });
+
+          applyRuleContexts({
+            helper,
+            ruleContexts,
+          });
+        }
+
+        render(
+          {
+            userData: [],
```

Comment:
It feels weird to have it called like that specially when `transformItems` is called on it.

Reply:
`transformItems` is only called on `render` for now.

Should we rather return this?

```js
render(
  {
    userData: transformItems([]),
    instantSearchInstance,
    widgetParams,
  },
  true
);
```

Reply:
Actually I wasn't talking about that but more about the usage of the API. It's not natural to use `transformItems` on something that at the end is called `userData`. But I get the point that since query rules is a larger spectrum maybe we'll have `items` at some point.